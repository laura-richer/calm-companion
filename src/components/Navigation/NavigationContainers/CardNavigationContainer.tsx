import React from 'react';
import { View } from 'react-native';
import { StackNavigationProp } from '@react-navigation/stack';
import { useNavigation, ParamListBase } from '@react-navigation/native';
import { useDispatch } from 'react-redux';
import { useAppSelector } from '~store/hooks';
import { setActiveScreen } from '~store/slices/navigation';
import { NavigationItemType } from '~types/navigation';
import styles from '../styles';
import CardItem from '../subcomponents/CardItem';

const CardNavigationContainer = () => {
  const dispatch = useDispatch();
  const navigation = useNavigation<StackNavigationProp<ParamListBase>>();
  const navigationItems = useAppSelector(state => state.navigation.items);

  const handleChangeScreen = (id: string) => {
    dispatch(setActiveScreen(id));
    navigation.navigate(id);
  };

  return (
    <View style={styles.cardNavigationContainer} accessibilityLabel="Navigation items">
      {navigationItems.map((item: NavigationItemType) => (
        <CardItem key={item.id} item={item} handleChangeScreen={handleChangeScreen} />
      ))}
    </View>
  );
};

export default CardNavigationContainer;
