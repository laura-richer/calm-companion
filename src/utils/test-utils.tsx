import React, { PropsWithChildren } from 'react';
import { NavigationContainer } from '@react-navigation/native';
import { render, RenderOptions } from '@testing-library/react-native';
import { Provider as StoreProvider } from 'react-redux';
import { EnhancedStore, configureStore } from '@reduxjs/toolkit';
import { SafeAreaProvider } from 'react-native-safe-area-context';
import { RootState } from '~store/index';
import navigationReducer, {
  initialState as navigationInitialState,
} from '~store/slices/navigation';
import quickHelpReducer, { initialState as quickHelpInitialState } from '~store/slices/quickHelp';

type DeepPartial<T> = {
  [K in keyof T]?: T[K] extends object ? DeepPartial<T[K]> : T[K];
};

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: Partial<RootState>;
  store?: EnhancedStore;
}

export const renderWithProviders = (
  ui: React.ReactElement,
  { preloadedState = {}, store = createTestStore(preloadedState) }: ExtendedRenderOptions = {}
) => {
  const wrapper = ({ children }: PropsWithChildren<{}>): JSX.Element => {
    return (
      <StoreProvider store={store}>
        <SafeAreaProvider>
          <NavigationContainer>{children}</NavigationContainer>
        </SafeAreaProvider>
      </StoreProvider>
    );
  };

  return {
    ...render(ui, {
      wrapper,
    }),
  };
};

const _createStore = (preloadedState: Partial<RootState> = {}): EnhancedStore => {
  return configureStore({
    reducer: {
      navigation: navigationReducer,
      quickHelp: quickHelpReducer,
    },
    preloadedState,
    middleware: getDefaultMiddleware =>
      getDefaultMiddleware({
        immutableCheck: { warnAfter: 256 },
        serializableCheck: { warnAfter: 256 },
      }),
  });
};

export const createTestStore = (preloadedState: DeepPartial<RootState> = {}): EnhancedStore => {
  const state: RootState = {
    // @ts-ignore
    navigation: {
      ...navigationInitialState,
      ...(preloadedState.navigation || {}),
    },
    // @ts-ignore
    quickHelp: {
      ...quickHelpInitialState,
      ...(preloadedState.quickHelp || {}),
    },
  };
  const store: EnhancedStore = _createStore(state);
  const origDispatch = store.dispatch;
  store.dispatch = jest.fn(origDispatch);
  return store;
};
