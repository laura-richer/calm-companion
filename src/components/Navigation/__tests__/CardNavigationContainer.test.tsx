import React from 'react';
import { screen } from '@testing-library/react-native';
import { renderWithProviders } from '~utils/test-utils';
import CardNavigationContainer from '../NavigationContainers/CardNavigationContainer';

// const mockNavigate = jest.fn();

describe('Card navigation container', () => {
  beforeEach(() => {
    renderWithProviders(<CardNavigationContainer />);
  });

  test('matches snapshot', () => {
    expect(screen.toJSON()).toMatchSnapshot();
  });

  test('renders quick help button', () => {
    const tab = screen.getByLabelText('Quick Help');
    expect(tab).toBeOnTheScreen();
  });

  test('renders anxiety resources tab', () => {
    const tab = screen.getByLabelText('Anxiety Resources');
    expect(tab).toBeOnTheScreen();
  });

  test('renders cbt tools tab', () => {
    const tab = screen.getByLabelText('CBT Tools');
    expect(tab).toBeOnTheScreen();
  });

  test('renders helpful contacts tab', () => {
    const tab = screen.getByLabelText('Helpful Contacts');
    expect(tab).toBeOnTheScreen();
  });

  // test('navigates to screen on press', async () => {
  //   const button = screen.getByLabelText('Quick Help');

  //   await fireEvent.press(button);

  //   expect(mockNavigate).toHaveBeenCalledWith('QuickHelp');
  // });
});
