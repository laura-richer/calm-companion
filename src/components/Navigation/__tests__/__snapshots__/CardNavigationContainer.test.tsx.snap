// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card navigation container matches snapshot 1`] = `
<View
  accessibilityLabel="Navigation items"
  style={
    {
      "display": "flex",
      "flexDirection": "row",
      "flexWrap": "wrap",
      "justifyContent": "space-between",
    }
  }
>
  <View
    accessibilityLabel="Quick Help"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "borderRadius": 8,
          "flexBasis": "47%",
          "marginBottom": 20,
          "padding": 10,
        },
        {
          "shadowColor": "#8b8490",
          "shadowOffset": {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.25,
          "shadowRadius": 8,
        },
        {
          "backgroundColor": "#ffffff",
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "fontFamily": "Poppins-Bold",
            "fontSize": 14,
          },
          {
            "color": "#3f7b82",
            "maxWidth": 100,
          },
        ]
      }
    >
      Quick
      

      Help
    </Text>
    <Image
      accessibilityHint="Icon"
      alt="Quick Help icon"
      resizeMode="contain"
      source={
        {
          "testUri": "../../../src/assets/images/screenHeaders/quick-help.png",
        }
      }
      style={
        {
          "marginLeft": "auto",
          "maxWidth": 75,
          "opacity": 0.25,
        }
      }
    />
  </View>
  <View
    accessibilityLabel="Anxiety Resources"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "borderRadius": 8,
          "flexBasis": "47%",
          "marginBottom": 20,
          "padding": 10,
        },
        {
          "shadowColor": "#8b8490",
          "shadowOffset": {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.25,
          "shadowRadius": 8,
        },
        {
          "backgroundColor": "#ffffff",
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "fontFamily": "Poppins-Bold",
            "fontSize": 14,
          },
          {
            "color": "#3f7b82",
            "maxWidth": 100,
          },
        ]
      }
    >
      Anxiety
      

      Resources
    </Text>
    <Image
      accessibilityHint="Icon"
      alt="Anxiety Resources icon"
      resizeMode="contain"
      source={
        {
          "testUri": "../../../src/assets/images/screenHeaders/anxiety-resources.png",
        }
      }
      style={
        {
          "marginLeft": "auto",
          "maxWidth": 75,
          "opacity": 0.25,
        }
      }
    />
  </View>
  <View
    accessibilityLabel="CBT Tools"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "borderRadius": 8,
          "flexBasis": "47%",
          "marginBottom": 20,
          "padding": 10,
        },
        {
          "shadowColor": "#8b8490",
          "shadowOffset": {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.25,
          "shadowRadius": 8,
        },
        {
          "backgroundColor": "#ffffff",
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "fontFamily": "Poppins-Bold",
            "fontSize": 14,
          },
          {
            "color": "#3f7b82",
            "maxWidth": 100,
          },
        ]
      }
    >
      CBT
      

      Tools
    </Text>
    <Image
      accessibilityHint="Icon"
      alt="CBT Tools icon"
      resizeMode="contain"
      source={
        {
          "testUri": "../../../src/assets/images/screenHeaders/cbt-tools.png",
        }
      }
      style={
        {
          "marginLeft": "auto",
          "maxWidth": 75,
          "opacity": 0.25,
        }
      }
    />
  </View>
  <View
    accessibilityLabel="Helpful Contacts"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "borderRadius": 8,
          "flexBasis": "47%",
          "marginBottom": 20,
          "padding": 10,
        },
        {
          "shadowColor": "#8b8490",
          "shadowOffset": {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.25,
          "shadowRadius": 8,
        },
        {
          "backgroundColor": "#ffffff",
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "fontFamily": "Poppins-Bold",
            "fontSize": 14,
          },
          {
            "color": "#3f7b82",
            "maxWidth": 100,
          },
        ]
      }
    >
      Helpful
      

      Contacts
    </Text>
    <Image
      accessibilityHint="Icon"
      alt="Helpful Contacts icon"
      resizeMode="contain"
      source={
        {
          "testUri": "../../../src/assets/images/screenHeaders/helpful-contacts.png",
        }
      }
      style={
        {
          "marginLeft": "auto",
          "maxWidth": 75,
          "opacity": 0.25,
        }
      }
    />
  </View>
</View>
`;
